Initial Examples Page

kurtmilam opened the pull request:
Plus related changes to routes and header

kurtmilam committed:
Initial Examples Page

Plus related changes to routes and header

kurtmilam committed:
Added makeUrl function to querystring-params.js

kurtmilam committed:
Deleted unnecessary log call

polytypic commented:
I took a very brief glance of over the PR and it seems quite nice!  I have a couple of things in mind I'd recommend doing differently, but nothing major.  I'll write notes on those later&mdash;I need to do other things right now.  Don't let that delay you from working on. 👍

kurtmilam committed:
Split getter and setter functions out of subPathL

Also improved the pathRoot regex and suppressed the ability to add a
forward slash at the beginning of the PathInput.

kurtmilam committed:
Renamed makeUrl to makePath

kurtmilam committed:
Call PrettyStringify like a component, encode querystring keys & values

kurtmilam committed:
Ditch makePath for makeQuerystring

kurtmilam committed:
Added decodeProps function

Decode pathParam values prior to stringifying them.

kurtmilam committed:
Added newPathString fn, moved paramsPairs out of QueryStringParams

kurtmilam committed:
Import restricted-input.less, add pretty-stringify CSS class

Changed the CSS selector for pretty-stringify from element pre to class
pretty-stringify.

polytypic commented on src/public/components/pretty-stringify.js:
```diff
@@ -0,0 +1,7 @@
+export const PrettyStringify = ( { space, obj } ) =>
+  <pre className="pretty-stringify">
+    { U.lift( x => JSON.stringify( x, null, Number( space ) ) )( obj ) }
+  </pre>
```
I was actually previously going to suggest turning this into a proper component taking props. 👍

Note that written this way the `space` needs to be a (numeric) constant, which is something to watch out.  Also, when there is just a single component exported from a module I tend to export it as the default.

How about something like this:

```js
export const PrettyStringify = ( { value, replacer = null, space = null } ) =>
   <pre className="pretty-stringify">
     { U.lift(JSON.stringify)(value, replacer, space) }
   </pre>
```

This just kind of directly wraps `JSON.stringify` as a component.  All the parameters are supported with the normal defaults.  Perhaps `space` should be defaulted to `2` instead?  I took the prop names from the parameter names here: https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify

BTW, I find this idea of wrapping `JSON.stringify` neat.

kurtmilam replied:
Agreed. I'll add your suggested changes and default `space` to 2. Should I force the default if `space` `isNaN`?

polytypic replied:
I usually try to avoid having too much intelligence in program components (in general, not just UI components) and treat cases like that as bugs.  IOW, it would be responsibility of the user to make sure that `space` parameter is valid.

kurtmilam replied:
How about this:
```js
export default ( { value, replacer = null, space } ) =>
  <pre className="pretty-stringify">
    { U.lift( JSON.stringify )( value, replacer, U.ifElse( isNaN, _ => 2, Number )( space ) ) }
  </pre>
```
This works whether I pass in `space="2"`, `space={ 2 }`, `space={ U.atom( 2 ) }`, or NaN.

polytypic commented on src/public/pages/examples/path-params.js:
```diff
@@ -0,0 +1,42 @@
+      type="text"
+      label="Path"
+      value={ U.view( subPathL( getPagePathRoot( path ) ), path ) }/>
+    <PrettyStringify space="2" obj={ decodeProps( props ) } />
```
Note that you can write `space={2}` to pass number `2` instead of string `"2"` as a prop.

kurtmilam replied:
Yes, I've variously passed in `"2"` and `{ 2 }`. I thought it might make sense to accept either one?

polytypic replied:
Hmm... Usually I try to avoid implicit conversions like that, but this is definitely  case where the implicit conversion is unlikely to cause major problems.  One thing that implicit conversions like that can hide is bugs where you are just passing something completely wrong to the component like accidentally passing the wrong variable from scope. What would happen then is that it is converted to NaN and then treated as default and that can hide the bug (potentially for a long time).

kurtmilam replied:
I completely agree. Implicit conversions are an absolute rarity in my code. I thought this might be an OK place to relax that rule, but on second thought, it's probably better to stick to principles in code written for a boilerplate / examples project. Don't want to give an implicit green light to bad practices.

I've gone back to the following:
```js
export default ( { value, replacer = null, space = 2 } ) =>
  <pre className="pretty-stringify">
    { U.lift( JSON.stringify )( value, replacer, space ) }
  </pre>
```
It now works with (and only with) `undefined`, `numericConstant` or `U.atom( numericConstant )`.

kurtmilam commented:
Side note: I definitely like the path-params component better than the querystring-params component.

polytypic commented on src/public/components/text-input.js:
```diff
@@ -0,0 +1,7 @@
+export const TextInput = ( { value } ) =>
+  <input type="text"
+         value={ value }
+         onChange={ U.getProps( { value } ) }/>
```
The example in this note is not something that should be done here, but here is an example of a couple patterns I've often used with components that are basically just simple wrappers for HTML elements:

```js
export const TextInput = ( { value, mount, onChange, ...props } ) =>
   <input type="text"
          value={ value }
          onChange={ U.actions( U.getProps( { value } ), onChange ) }
          ref={ mount }
          {...props}/>
```

* This gives the possibility to customize the component by giving it additional props like `className`.
* It is also possible to give it a `mount` property that is translated to `ref` to get access to the DOM `input` element.
* The outer `onChange` is composed with the built-in action.  Sometimes similar patterns are useful for merging `className`s given as parameters.

kurtmilam replied:
This is a very helpful example that answers some questions I had. I had considered adding `...props` to some of the components - I had it on `pretty-stringify`, then decided to remove it. The notes about `mount` and `onChange` are new and helpful information for me, as well.

kurtmilam replied:
Would it make sense to have most components accept `mount` and `...props`?

Also, do you have an example that shows how to work with the `ref` translated from a `mount`? I like to set focus on inputs intelligently / automatically (so the user doesn't have to tab through or use their mouse to put focus where you're pretty sure they'd want to have it after some action). For instance, in `QuerystringParams`, I think it would make sense to set focus to the 'key' input of a newly added key/value pair after the user clicks the 'Add' button.

polytypic commented on src/public/pages/examples/querystring-params.js:
```diff
@@ -0,0 +1,63 @@
+          )
+
+const paramsPairs = params =>
+  U.view( L.iso( U.toPairs, U.fromPairs ), params )
```
The `params` parameter can be eliminated here as `U.view` is curried.

BTW, it can be advantageous to keep accessors like this as plain optics and call `U.view` only at point of where you have data (params) at hand.  The reason for that is that `U.view(lens, data)` returns a view of the `data`, which is not as composable as the `lens`.  In a context like this the additional composability isn't needed, of course, but in the general case that is good thing to keep in mind.

polytypic commented on src/public/pages/examples/querystring-params.js:
```diff
@@ -0,0 +1,63 @@
+      <tbody>
+        { U.mapElems( ( param, i ) =>
+                        <QuerystringParam key={ i } param={ param }/>
+                    )( copy )
```
Hmm... Note that you can directly pass both parameters: `U.mapElems(... => ..., copy)`.  Doing so is actually a tiny bit faster and also a bit more concise.

polytypic commented on src/public/pages/examples/querystring-params.js:
```diff
@@ -0,0 +1,63 @@
+    </table>
+    Navigate to:&nbsp;
+    <Link href={ newPathString( path, copy ) }>
+      { newPathString( path, copy ) }
```
This is obvious, of course, but here the property `newPathString( path, copy )` is computed twice.  This gives you two independent copies of the resulting observable and both of them compute the path string independently after each change.  It doesn't really matter in this case, but it can be better to bind such computations to a variable and use the variable twice to avoid duplicated computation.

BTW, Karet Util has a `scope` function, which can be convenient:

```js
{U.scope((href = newPathString(path, copy)) =>
   <Link href={href}>{href}</Link> )}
```

`U.scope(fn)` is equivalent to `fn()`.

Or you could have a link component that just displays the `href`.

polytypic commented on src/public/pages/examples/querystring-params.js:
```diff
@@ -0,0 +1,63 @@
+const QuerystringParam = ( { param } ) =>
+  <tr>
+    <td><TextInput value={ U.view( [ 0 ], param ) }/></td>
+    <td><TextInput value={ U.view( [ 1 ], param ) }/></td>
```
Here I would suggest dropping the (unnecessary) array:  `U.view( 0, param )`.

polytypic commented on src/public/pages/examples/path-params.js:
```diff
@@ -0,0 +1,42 @@
+  L.iso( subPathGetter( pagePathRoot )
+       , subPathSetter( pagePathRoot )
+       )
+const subPathL = U.lift( _subPathL )
```
Hmm... Perhaps just drop `_subPathL` and use `U.lift`:

```js
const subPathL = U.lift( pagePathRoot =>
  L.iso( subPathGetter( pagePathRoot )
       , subPathSetter( pagePathRoot )
       )
  )
```

polytypic commented on src/public/components/restricted-input.js:
```diff
@@ -0,0 +1,43 @@
+import React     from "karet"
+
+export const RestrictedInput = ({value, meta: {format, parse}, ...props}) => {
+  const edited = U.atom(undefined)
```
This comes from the examples repo, but nowadays I generally recommend exposing (by default) local state like this as parameters.

kurtmilam replied:
I think I understand. You're saying you'd pass in `edited` so the caller could also be aware of changes to `edited` in case it needed to be, correct? Would you pass anything else in like `shown` and/or `validity`? And where would you put them in the args list? i.e. `({value, meta: {format, parse, edited}, ...props})` or what?

polytypic replied:
I mean that the local `edited` variable would simply be exposed as a parameter.  This allows a user of the `RestrictedInput` to store the `edited` state outside of the component.  I would still recommend giving `edited` a default value (making it local state, unless the user decides otherwise), because in most cases that will work just fine.  IOW:

```js
export const RestrictedInput = ({value, meta: {format, parse}, edited = U.atom(), ...props}) => {
```

polytypic commented:
Overall this looks very nice!

kurtmilam commented:
Thanks for the thorough review! The example with `U.scope` was particularly helpful. I hadn't understood exactly what it was for until you explained it in the context of this code.

I don't know about you, but I'm inclined to go through and implement all of your suggestions for several reasons. Selfishly, it'll be good practice for me. Additionally, it will help propagate best practices and patterns to anyone who decides to play with the boilerplate. And, lastly, it will mean the boilerplate is delivered with a few useful components that someone could continue using in a project based on the boilerplate.

What do you think?

kurtmilam committed:
Incorporate a number of suggestions from @polytypic

kurtmilam commented:
A couple of questions on `QuerystringParams`:
1. I was having a slightly difficult time translating your side-effect-less patterns from a string to an array of pairs. If you find some time, could you help me out with that?
2. I get `Uncaught TypeError: Cannot read property 'length' of undefined` when I try to remove the last remaining key/value pair. I tried a number of things to make the error go away without any luck.

polytypic commented:
Regarding 2, sounds like `L.define([])` needs to be inserted at an appropriate place to avoid partial lenses from eliminating the array.

kurtmilam commented:
Believe it or not, I had tried inserting `L.define( [] )` on my own, yesterday, but hadn't yet figured out where it needed to go. Now that I know that is probably a part of the correct solution, I'll go back through and see whether I can figure out where to insert it.

polytypic commented:
```js
export const QuerystringParams = ( { params, path, copy = U.atom() } ) => {
  const copied =
    U.view(K(params, params =>
             [L.defaults(params),
              toFromPairsL,
              L.define([])]),
           copy)

  return (
    <div>
      <table>
        <thead>
          <tr>
            <td>Key</td>
            <td>Value</td>
            <td>
              <button onClick={ () => addQuerystringParams( copied ) }>Add</button>
            </td>
          </tr>
        </thead>
        <tbody>
          { U.mapElems( ( param, i ) =>
                          <QuerystringParam key={ i } param={ param }/>
                        , copied
                      )
          }
        </tbody>
      </table>
      Navigate to:&nbsp;
      { U.scope( ( href = newPathString( path, copied ) ) =>
                   <Link href={ href }>{ href }</Link>
               )
      }
      <PrettyStringify value={ params } />
    </div>
  )
}
```

kurtmilam commented:
A-ha! That makes sense, thanks!

kurtmilam committed:
Fixed undefined error

Implemented @polytypic's suggestion for inserting L.define([]) to avoid
the error that came from trying to delete the last key/value pair.

kurtmilam committed:
Make 'edited' an optional argument

Following @polytypic's suggestion. This makes it possible for the caller
to know the component's edited state, in case it's interested in knowing
that.

kurtmilam commented:
I think I've implemented most of your suggestions, although there are probably a few components I could change to take `mount` and `...props`. I'll probably hit those at some other time, since I have some other things to work on this afternoon.

polytypic commented:
Hmm... I believe it is better to avoid having too many extra features in samples in general.  I believe that easily leads people casually looking at the examples to think that the extra things are necessary.  So, I believe having examples with `...props` in a place or two, for reference, is better that having it everywhere.

kurtmilam committed:
Code format and changed querystring check from isNil to equals([])

polytypic commented:
Ready to merge?

kurtmilam commented:
Ready!

polytypic commented:
Thanks!
